### PR TITLE
[added] highlighting for useful class items on [C]

### PR DESCRIPTION
**Pull request by Diprog**
animation

Performance with tons of gold:
```[12:05:07] [0.039s total, 0.00004s avg]     1056 calls - void onTick(CBlob@ this):warmusic.as
[12:05:07] [0.039s total, 0.00004s avg]     1056 calls - void onTick(CRules@ this):spawnfish.as
[12:05:07] [0.041s total, 0.00004s avg]      991 calls - void onRender(CMap@ this):sandbox.as
[12:05:07] [0.044s total, 0.00004s avg]     1056 calls - void onTick(CBlob@ this):blockplacement.as
[12:05:07] [0.049s total, 0.00005s avg]     1056 calls - void onTick(CSprite@ sprite):highlightitems.as
[12:05:07] [0.056s total, 0.00005s avg]     1056 calls - void onTick(CBlob@ this):standardcontrols.as
[12:05:07] [0.056s total, 0.00005s avg]     1056 calls - void onTick(CBlob@ this):builderlogic.as
[12:05:07] [0.061s total, 0.00006s avg]     1056 calls - void onTick(CSprite@ this):builderanim.as
[12:05:07] [0.063s total, 0.00006s avg]     1056 calls - void onTick(CMovement@ this):runnermovement.as
[12:05:07] [0.093s total, 0.00009s avg]     1056 calls - void onTick(CSprite@ this):runnerhead.as
[12:05:07] [0.120s total, 0.00012s avg]      991 calls - void onRender(CRules@ this):markplayers.as
[12:05:07] [0.192s total, 0.00018s avg]     1056 calls - void onTick(CBlob@ this):drawhovermessages.as
[12:05:07] [0.551s total, 0.00056s avg]      991 calls - void onRender(CSprite@ this):builderhud.as
[12:05:07] [0.556s total, 0.00056s avg]      991 calls - void onRender(CSprite@ this):defaultactorhud.as
[12:05:07] [1.591s total, 0.00161s avg]      991 calls - void onRender(CSprite@ sprite):highlightitems.as
[12:05:07] [2.022s total, 0.00204s avg]      991 calls - void onRender(CSprite@ this):standardpickup.as
[12:05:07] [3.747s total, 0.00355s avg]     1056 calls - void onTick(CRules@ this):mechanismsserver.as
[12:05:07] [13.707s total, 0.01298s avg]     1056 calls - void onTick(CBlob@ this):standardpickup.as```

### CODE REVIEW

File: Entities/Common/Controls/HighlightItemsCommon.as [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	{
+		if (classes[i] == config_name)
+		{
+			return i;
+		}
+	}
+	return -1;
+}
+
+bool isItemToHighlight(CBlob@ for_blob, CBlob@ item)
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItemsCommon.as]: `shouldHighlightBlob` seems better

File: Entities/Common/Controls/HighlightItemsCommon.as [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+{
+	string item_config_name = item.getConfig();
+	int class_index = getClassIndex(for_blob);
+	for (uint i = 0; i < highlight_items[class_index].length(); i++)
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItemsCommon.as]: use `array.find`

File: Entities/Common/Controls/HighlightItemsCommon.as (hunk 1 above)
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItemsCommon.as]: most logic should be merged with the below overload

File: Entities/Common/Controls/StandardPickup.as [hunk 3]
```diff
@@ -405,10 +406,11 @@ void onRender(CSprite@ this)
 				CBlob @b = pickupBlobs[i];
 
 				bool canBePicked = canBlobBePickedUp(blob, b);
-
+				bool is_item_to_highlight = isItemToHighlight(blob, b);
```
Comment (AsuMagic) [inline Entities/Common/Controls/StandardPickup.as]: variable should be named `shouldHighlight`

File: Entities/Common/Controls/HighlightItems.as [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+			show_names = true;
+		else
+			show_names = false;
+	}
+}
+
+void onRender(CSprite@ this)
+{
+	CBlob@ blob = this.getBlob();
+	if (blob !is null && show_names)
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItems.as]: can `CSprite.getBlob()` ever be `null`?

File: Entities/Common/Controls/HighlightItems.as [hunk 5]
```diff
@@ -0,0 +1,81 @@
+#define CLIENT_ONLY
+#include "HighlightItemsCommon.as"
+bool show_names = false;
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItems.as]: `show_names` does not seem accurate

File: Entities/Common/Controls/HighlightItems.as [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+//Items to highlight.
+const string[] classes = {"builder", "knight", "archer"};
+const string[][] highlight_items = {
+/* 0 */	{"mat_stone", "mat_wood", "mat_gold"}, //builder
+/* 1 */	{"mat_bombs", "mat_waterbombs"}, //knight
+/* 2 */	{"mat_firearrows", "mat_waterarrows", "mat_bombarrows"} //archer
+};
+
+//If button is pressed.
+bool do_highlight = false;
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItems.as]: remove this variable and perform the check in `onRender`

File: Entities/Common/Controls/HighlightItems.as [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+{
+	if (do_highlight)
+	{
+		//Highlight items.
+		CCamera@ camera = getCamera();
+		CMap@ map = getMap();
+		if (camera is null || map is null) return;
+
+		CBlob@[] blobs;
+		if (getBlobs(@blobs))
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItems.as]: use `CMap.getBlobsInBox` to get only visible blobs

File: Entities/Common/Controls/HighlightItems.as [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+		{
+			for (uint i = 0; i < blobs.length; i++)
+			{
+				CBlob@ blob = blobs[i];
+				//Highlight.
+				int class_index = classes.find(sprite.getBlob().getConfig());
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItems.as]: shouldn't it be `.getName()`?

File: Entities/Common/Controls/HighlightItems.as [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+				if (shouldHighlightBlob(class_index, blob) && !blob.isInInventory())
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItems.as]: check for `isInInventory()` before checking for more complicated logic

File: Entities/Common/Controls/HighlightItems.as [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+				{
+					float luminance = map.getColorLight(blob.getPosition()).getLuminance();
+					//Don't highlight items in dark caves and places.
+					if (luminance >= hide_luminance_level)
+					{
+						//Highlight like any normal pickup.
+						blob.RenderForHUD(Vec2f_zero, 0.0f, SColor(255,255,255,255), RenderStyle::normal);
+						//But do a beautiful fading effect.
+						uint brightness_level = Maths::Abs(Maths::Sin(getGameTime() / 20.0f) * 180);
```
Comment (AsuMagic) [inline Entities/Common/Controls/HighlightItems.as]: I would try to keep the period in seconds separate, as in, make the multiplication by 2pi explicit

**Comment by AsuMagic** [comment]
Latest image since diprog didn't send it there:

latest image

**Comment by 1bardesign** [comment]
As this is somewhat contentious I'm going to leave it out for now. However, I'm not opposed to the core idea of the change and helping people find their items. Maybe we can find an implementation that everyone is happy with?

Sidenote; having discussion and review split between forums and github isn't great.

**Comment by AsuMagic** [comment]
I would keep the code review discussion on github and the idea discussion on forums.

**Comment by AsuMagic** [comment]
(not mergeable atm, he accidentally removed isMyPlayer checks which may be bad, and he's afk now)

**Comment by 1bardesign** [comment]
Added some discussion [here](https://forum.thd.vg/threads/highlight-useful-items-for-your-class-on-c.27408/page-2#post-404508) about everything we need to do and decide before this gets merged. I don't want this to just become stale and everyone be dissatisfied with what happened, but there's also a few things that need to be implemented and talked about before i'm ok merging this too.

**Comment by 1bardesign** [comment]
I'm leaning towards merging this after asu's "60fps" work is through testing - I'd rather we merge optimistically than pessimistically, and I'd like to have this closed one way or another within the month as it's our longest-outstanding PR.

Any opposition - please make your case here.

**Comment by benjamingorman** [comment]
@Diprog how did you get all the benchmarks for this?

**Comment by AsuMagic** [comment]
It's [the new profiler](https://forum.thd.vg/threads/new-profiler-in-build-2409.27459/).

**Comment by benjamingorman** [comment]
Wow I totally missed this. Seems super useful.

**Comment by 1bardesign** [comment]
Going ahead with this - we'll tweak it in test build and based on public feedback surely, but it seems useful.

**Pull request merged**